hir2/src/patterns/pattern.rs

```diff
@@ -0,0 +1,382 @@
+            let op_name = <Shl as crate::OpRegistration>::register_with(&*dialect);
+            let mut info = PatternInfo::new(
+                context,
+                "convert-shl1-to-mul2",
+                PatternKind::Operation(op_name),
+                PatternBenefit::new(1),
+            );
+            info.with_bounded_rewrite_recursion(true);
+            Self { info }
+        }
+    }
+    impl Pattern for ConvertShiftLeftBy1ToMultiply {
+        fn info(&self) -> &PatternInfo {
+            &self.info
+        }
+    }
+    impl RewritePattern for ConvertShiftLeftBy1ToMultiply {
+        fn matches(&self, op: OperationRef) -> Result<bool, Report> {
+            use crate::matchers::{self, match_chain, match_op, MatchWith, Matcher};
+
+            let binder = MatchWith(|op: &UnsafeIntrusiveEntityRef<Shl>| {
+                log::trace!(
+                    "found matching 'hir.shl' operation, checking if `shift` operand is foldable"
+                );
+                let op = op.borrow();
+                let shift = op.shift().as_operand_ref();
+                let matched = matchers::foldable_operand_of::<Immediate>().matches(&shift);
+                matched.and_then(|imm| {
+                    log::trace!("`shift` operand is an immediate: {imm}");
+                    let imm = imm.as_u64();
+                    if imm.is_none() {
+                        log::trace!("`shift` operand is not a valid u64 value");
+                    }
+                    if imm.is_some_and(|imm| imm == 1) {
+                        Some(())
+                    } else {
+                        None
+                    }
+                })
+            });
+            log::trace!("attempting to match '{}'", self.name());
+            let matched = match_chain(match_op::<Shl>(), binder).matches(&op.borrow()).is_some();
+            log::trace!("'{}' matched: {matched}", self.name());
+            Ok(matched)
+        }
+
+        fn rewrite(&self, op: OperationRef, rewriter: &mut dyn Rewriter) {
+            log::trace!("found match, rewriting '{}'", op.borrow().name());
+            let (span, lhs) = {
+                let shl = op.borrow();
+                let shl = shl.downcast_ref::<Shl>().unwrap();
+                let span = shl.span();
+                let lhs = shl.lhs().as_value_ref();
+                (span, lhs)
+            };
+            let constant_builder = rewriter.create::<Constant, _>(span);
+            let constant: UnsafeIntrusiveEntityRef<Constant> =
+                constant_builder(Immediate::U32(2)).unwrap();
+            let shift = constant.borrow().result().as_value_ref();
+            let mul_builder = rewriter.create::<Mul, _>(span);
+            let mul = mul_builder(lhs, shift, Overflow::Wrapping).unwrap();
+            let mul = mul.borrow().as_operation().as_operation_ref();
+            log::trace!("replacing shl with mul");
+            rewriter.replace_op(op, mul);
+        }
```

Comment:
Ergonomics look great! The code gaves strong MLIR vibes! It'd take some time to getting use to the types, techniques, etc. but it worth it. I'm looking forward to start writing some rewrites.